token/core/common/loaders.go

```diff
@@ -0,0 +1,223 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package common
+
+import (
+	"time"
+
+	"github.com/hyperledger-labs/fabric-smart-client/platform/view/view"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/driver"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/token"
+	"github.com/pkg/errors"
+)
+
+type LedgerToken interface {
+	GetOwner() []byte
+}
+
+type TokenDeserializer[T any] interface {
+	DeserializeToken([]byte) (T, error)
+}
+
+type MetadataDeserializer[M any] interface {
+	DeserializeMetadata([]byte) (M, error)
+}
+
+type TokenAndMetadataDeserializer[T LedgerToken, M any] interface {
+	TokenDeserializer[T]
+	MetadataDeserializer[M]
+}
+
+type VaultLedgerTokenLoader[T any] struct {
+	TokenVault   TokenVault
+	Deserializer TokenDeserializer[T]
+
+	// Variables used to control retry condition
+	NumRetries int
+	RetryDelay time.Duration
+}
+
+func NewLedgerTokenLoader[T any](tokenVault TokenVault, deserializer TokenDeserializer[T]) *VaultLedgerTokenLoader[T] {
+	return &VaultLedgerTokenLoader[T]{
+		TokenVault:   tokenVault,
+		Deserializer: deserializer,
+		NumRetries:   3,
+		RetryDelay:   3 * time.Second,
+	}
+}
+
+// GetTokenOutputs takes an array of token identifiers (txID, index) and returns the corresponding token outputs
+func (s *VaultLedgerTokenLoader[T]) GetTokenOutputs(ids []*token.ID) ([]T, error) {
+
+	var err error
+	for i := 0; i < s.NumRetries; i++ {
+		tokens := make([]T, len(ids))
+		counter := 0
+		err = s.TokenVault.GetTokenOutputs(ids, func(id *token.ID, bytes []byte) error {
+			if len(bytes) == 0 {
+				return errors.Errorf("failed getting serialized token output for id [%v], nil value", id)
+			}
+			ti, err := s.Deserializer.DeserializeToken(bytes)
+			if err != nil {
+				return errors.Wrapf(err, "failed deserializing token for id [%v][%s]", id, string(bytes))
+			}
+			tokens[counter] = ti
+			counter++
+			return nil
+		})
+		if err == nil {
+			return tokens, nil
+		}
+
+		// check if there is any token id whose corresponding transaction is pending
+		// if there is, then wait a bit and retry to load the outputs
+		retry := false
+		for _, id := range ids {
+			pending, err := s.TokenVault.IsPending(id)
```

Comment:
Maybe we could do a batch request here by changing the signature of `TokenVault`
```
IsPending(ids ...TokenID) map[TokenID]bool
```
or at least add a todo

Reply:
👍